[PN-68] Refactor  _robot_command_goal_complete

abaker-bdai opened the pull request:
**Merge https://github.com/bdaiinstitute/spot_ros2/pull/179 before merging this PR.**

Adds unit test for _`robot_command_goal_complete` and refactors the function so that it is easier to read.

While writing the test, I spotted a few statuses that potentially return the incorrect value. I'll implement some of these test cases in a follow up PR.

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
+
+        # Should this status return a GoalResponse.FAILED?
+        feedback.feedback.arm_cartesian_feedback.status.value = (
+            feedback.feedback.arm_cartesian_feedback.status.STATUS_TRAJECTORY_CANCELLED
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
+        # Should this status return a GoalResponse.FAILED?
+        feedback.feedback.arm_cartesian_feedback.status.value = (
+            feedback.feedback.arm_cartesian_feedback.status.STATUS_TRAJECTORY_STALLED
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
```
Yes I think in both these cases we should be returning failed.

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
+
+        # Should this status return a GoalResponse.FAILED?
+        feedback.feedback.arm_joint_move_feedback.status.value = (
+            feedback.feedback.arm_joint_move_feedback.status.STATUS_STALLED
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
```
Yes I think this should be failed

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
+
+        # Should this status return a GoalResponse.FAILED?
+        feedback.feedback.named_arm_position_feedback.status.value = (
+            feedback.feedback.named_arm_position_feedback.status.STATUS_STALLED_HOLDING_ITEM
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
```
Yes should be failed

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
+
+        """ Testing arm velocity feedback """
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_ARM_VELOCITY_FEEDBACK_SET
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), None)
```
Are we really supposed to return None here?

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        feedback.feedback.arm_gaze_feedback.status.value = feedback.feedback.arm_gaze_feedback.status.STATUS_IN_PROGRESS
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
+
+        # Should this status return a GoalResponse.FAILED?
+        feedback.feedback.arm_gaze_feedback.status.value = (
+            feedback.feedback.arm_gaze_feedback.status.STATUS_TOOL_TRAJECTORY_STALLED
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.IN_PROGRESS)
```
Yes failed

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        feedback.feedback.arm_drag_feedback.status.value = feedback.feedback.arm_drag_feedback.status.STATUS_DRAGGING
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.SUCCESS)
+
+        feedback.feedback.arm_drag_feedback.status.value = feedback.feedback.arm_drag_feedback.status.STATUS_UNKNOWN
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.FAILED)
```
Should this be FAILED?  In all the other ones it's IN_PROGRESS

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.FAILED)
+
+        """ Testing arm drag feedback """
```
```suggestion
        """ Testing arm impedance feedback """
```

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        """ Testing arm drag feedback """
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_ARM_DRAG_FEEDBACK_SET
+
+        feedback.feedback.arm_drag_feedback.status.value = feedback.feedback.arm_drag_feedback.status.STATUS_DRAGGING
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), GoalResponse.SUCCESS)
```
Is this SUCCESS or IN_PROGRESS?  There doesn't seem to be any IN_PROGRESS return

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        """ Testing arm drag feedback """
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_ARM_IMPEDANCE_FEEDBACK_SET
+
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), None)
+
+        """ Testing arm Impedance feedback """
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_ARM_IMPEDANCE_FEEDBACK_SET
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), None)
+
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_NOT_SET
+
+        self.assertEqual(self.spot_ros2._process_synchronized_arm_command_feedback(feedback), None)
```
Hm returning None doesn't seem right...  Probably should be IN PROGRESS

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        feedback.status.value = feedback.status.STATUS_INCOMPATIBLE_HARDWARE
+        self.assertEqual(self.spot_ros2._process_synchronized_mobility_command_feedback(feedback), GoalResponse.FAILED)
+
+        feedback.status.value = feedback.status.STATUS_PROCESSING
```
Should there be a test under this?

abaker-bdai replied:
You have to set the status to `STATUS_PROCESSING` before evaluating commands so we don't need a test to directly follow setting this status value.

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        )
+
+        """ Testing se2 velocity feedback """
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_SE2_VELOCITY_FEEDBACK_SET
+        self.assertEqual(self.spot_ros2._process_synchronized_mobility_command_feedback(feedback), None)
```
Probably not right...

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        """ Testing stop feedback """
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_STOP_FEEDBACK_SET
+
+        self.assertEqual(self.spot_ros2._process_synchronized_mobility_command_feedback(feedback), None)
+
+        """ Testing Follow arm feedback """
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_FOLLOW_ARM_FEEDBACK_SET
+
+        self.assertEqual(self.spot_ros2._process_synchronized_mobility_command_feedback(feedback), None)
+
+        """ Testing feedback not set"""
+        feedback.feedback.feedback_choice = feedback.feedback.FEEDBACK_NOT_SET
+
+        self.assertEqual(self.spot_ros2._process_synchronized_mobility_command_feedback(feedback), None)
```
Probably all of these should not be None

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        feedback.command.claw_gripper_feedback.status.value = (
+            feedback.command.claw_gripper_feedback.status.STATUS_AT_GOAL
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_gripper_command_feedback(feedback), None)
```
Not SUCCESS?

jbarry-bdai commented on spot_driver/test/test_spot_driver.py:
```diff
@@ -126,6 +136,489 @@ def test_wrapped_commands(self) -> None:
+        feedback.command.claw_gripper_feedback.status.value = (
+            feedback.command.claw_gripper_feedback.status.STATUS_APPLYING_FORCE
+        )
+        self.assertEqual(self.spot_ros2._process_synchronized_gripper_command_feedback(feedback), None)
```
Yeah this isn't right...  Probably should be SUCCESS or IN PROGRESS

jbarry-bdai reviewed (changes requested):
I've just made comments on the test because I suggest doing this PR in three pieces:
1) Only commit the test to reflect the current behavior of the function (but not the changes to the function included in this PR).
2) Commit a refactor to the function that behaves exactly as the old function did - confirmed by the test committed in 1
3) Commit changes to the refactor that fix the bugs discovered by the test.

This PR combines 1 and 2 - I suggest splitting it into two PRs.

Of the comments I made here, only the ones relevant to the actual test file (one typo, one place where I think a test is missing) should be updated in this PR.  The others should wait for the third step where we fix the implementation.

Thanks for this!

abaker-bdai commented:
> I've just made comments on the test because I suggest doing this PR in three pieces:
> 
> 1. Only commit the test to reflect the current behavior of the function (but not the changes to the function included in this PR).
> 2. Commit a refactor to the function that behaves exactly as the old function did - confirmed by the test committed in 1
> 3. Commit changes to the refactor that fix the bugs discovered by the test.
> 
> This PR combines 1 and 2 - I suggest splitting it into two PRs.
> 
> Of the comments I made here, only the ones relevant to the actual test file (one typo, one place where I think a test is missing) should be updated in this PR. The others should wait for the third step where we fix the implementation.
> 
> Thanks for this!

So the idea is that if you return `None`, the synchronous commands will return `GoalResponse.SUCCESS` later on, as long as no other commands return `GoalResponse.IN_PROGRESS` or `GoalResponse.FAILED`. What I can do is revert the test refactor to the initial test that I wrote to capture the pre-refactor state of the function. In a follow up PR, I'll address the bugs and potentially refactor the tests so that they are more thorough.

jbarry-bdai commented:
I think we should commit the test first and separately from changes to spot ros2.

abaker-bdai commented:
@jbarry-bdai Now that #179 has been merged into this branch, this is ready for review/stamp.

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
-        if feedback is None:
-            # NOTE: it takes an iteration for the feedback to get set.
+    def _process_full_body_command_feedback(self, feedback: FullBodyCommandFeedback) -> GoalResponse:
+        if feedback.status.value != feedback.status.STATUS_PROCESSING:
```
This could probably use a comment - I stared at it for a bit before remembering that there's more than one status in the message...

amessing-bdai replied:
@abaker-bdai

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+            if fb.payload_estimation_feedback.status.value == fb.payload_estimation_feedback.status.STATUS_ERROR:
+                return GoalResponse.FAILED
+            return GoalResponse.IN_PROGRESS
+        elif choice == fb.FEEDBACK_CONSTRAINED_MANIPULATION_FEEDBACK_SET:
+            if (
+                fb.constrained_manipulation_feedback.status.value
+                == fb.constrained_manipulation_feedback.status.STATUS_RUNNING
```
Does constrained manipulation never return SUCCESS?

amessing-bdai replied:
@abaker-bdai

abaker-bdai replied:
https://github.com/bdaiinstitute/bosdyn_msgs/blob/main/msg/ConstrainedManipulationCommandFeedbackStatus.msg Looking at the defined status values, it appears that it shouldn't return success. The status values imply either that it's running or it has failed to me

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+        fb = feedback.feedback
+        choice = fb.feedback_choice
+
+        if choice == fb.FEEDBACK_ARM_CARTESIAN_FEEDBACK_SET:
```
Where is the success return for this on trajectory complete?

jbarry-bdai replied:
I realized later that this happens in robot command goal complete - that's a bit confusing when reading it.  Could this either be called something different to indicate it only returns on non-success or maybe just have robot command goal complete return whatever is returned here?

jbarry-bdai reviewed (commented):
Not all the way through but sending what I have so far.

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+                return GoalResponse.FAILED
+            if fb.named_arm_position_feedback.status.value != fb.named_arm_position_feedback.status.STATUS_COMPLETE:
+                return GoalResponse.IN_PROGRESS
+        elif choice == fb.FEEDBACK_ARM_VELOCITY_FEEDBACK_SET:
+            self.get_logger().warn("WARNING: ArmVelocityCommand provides no feedback")
```
Will this spam?

amessing-bdai replied:
It's shouldn't the command should just immediately return success

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+            if fb.arm_gaze_feedback.status.value != fb.arm_gaze_feedback.status.STATUS_TRAJECTORY_COMPLETE:
+                return GoalResponse.IN_PROGRESS
+        elif choice == fb.FEEDBACK_ARM_STOP_FEEDBACK_SET:
+            self.get_logger().warn("WARNING: Stop command provides no feedback")
```
Spam?

abaker-bdai replied:
I believe in these cases where no feedback is provided, `_robot_command_goal_complete` will return `GoalResponse.SUCCESS` and proceed to the next command. So it shouldn't spam unless we are receiving endless stop commands

abaker-bdai replied:
And the same should apply to the other commands that don't provide feedback. I could just make these comments

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+        elif choice == fb.FEEDBACK_ARM_STOP_FEEDBACK_SET:
+            self.get_logger().warn("WARNING: Stop command provides no feedback")
+        elif choice == fb.FEEDBACK_ARM_DRAG_FEEDBACK_SET:
+            if fb.arm_drag_feedback.status.value != fb.arm_drag_feedback.status.STATUS_DRAGGING:
+                return GoalResponse.FAILED
```
Does this return SUCCESS on `STATUS_DRAGGING` and if so I think it should probably be `IN_PROGRESS`?

abaker-bdai replied:
Dragging would return `SUCCESS`, I'll change that to return `IN_PROGRESS`

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+            if fb.arm_drag_feedback.status.value != fb.arm_drag_feedback.status.STATUS_DRAGGING:
+                return GoalResponse.FAILED
+        elif choice == fb.FEEDBACK_ARM_IMPEDANCE_FEEDBACK_SET:
+            self.get_logger().warn("WARNING: ArmImpedanceCommand provides no feedback")
```
spam?

amessing-bdai replied:
The impedance command now has feedback from the upgrade to 3.3.2. Can we add handling of this?

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+    def _process_synchronized_gripper_command_feedback(
+        self, feedback: GripperCommandFeedback
+    ) -> Optional[GoalResponse]:
+        if (
+            feedback.status.value == feedback.status.STATUS_COMMAND_OVERRIDDEN
+            or feedback.status.value == feedback.status.STATUS_COMMAND_TIMED_OUT
+            or feedback.status.value == feedback.status.STATUS_ROBOT_FROZEN
+            or feedback.status.value == feedback.status.STATUS_INCOMPATIBLE_HARDWARE
+        ):
+            return GoalResponse.FAILED
```
Should this if statement be moved to a function somewhere?

abaker-bdai replied:
Good call, I'll pull this out to make it its own function.

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
+        return None
 
-        elif feedback.command.command_choice == feedback.command.COMMAND_SYNCHRONIZED_FEEDBACK_SET:
+    def _robot_command_goal_complete(self, feedback: RobotCommandFeedback) -> GoalResponse:
```
Oh right now I remember multiple feedback commands can be set so that's why you can't just return from one sub-feedback call.  But I think maybe having the `_process` return SUCCESS instead of None on SUCCESS would still be a lot clearer...

jbarry-bdai commented on spot_driver/spot_driver/spot_ros2.py:
```diff
@@ -1466,245 +1470,216 @@ def handle_max_vel(self, request: SetVelocity.Request, response: SetVelocity.Res
             # IN_PROGRESS.  And if it makes it to the bottom of the function, then all components
             # must be satisfied, and it returns SUCCESS.
-            # One corner case to know about is that the commands that don't provide feedback, such
+            # One corner if choice == to know about is that the commands that don't provide feedback, such
```
I think this may have been an unintentional find/replace change?

abaker-bdai replied:
Definitely. Thank you for catching this

amessing-bdai replied:
@abaker-bdai

jbarry-bdai reviewed (commented):
I think my only substantive comment is the confusion over `_process` functions not returning success.

abaker-bdai commented:
@amessing-bdai so it would be easier to review, I added the new feedback statuses and updated the tests here: https://github.com/bdaiinstitute/spot_ros2/pull/204
